ext/app/pipe/components.js

```diff
@@ -0,0 +1,322 @@
+          overflow: hidden;
+        }
+      </style>
+    `;
+    popup.querySelector('button').addEventListener('click', () => popup.remove());
+
+    // Remove any existing popup.
+    document.querySelectorAll('#grist-viewer-popup').forEach((el) => el.remove());
+
+    // Add the popup to the end of the body.
+    document.body.appendChild(popup);
+  }
+
+  function previewInGristClickHandler(refElement) {
+    // Read all the settings from the element.
+    const name = refElement.getAttribute('data-name');
+    const singlePage = refElement.hasAttribute('data-single-page');
+    // Allow all settings for the URL in that order. Href accepts both format and is easier to use.
+    const href =
+      refElement.getAttribute('href') ||
+      refElement.getAttribute('data-initial-file') || refElement.getAttribute('data-grist-doc-open') ||
+      refElement.getAttribute('data-initial-data') || refElement.getAttribute('data-grist-csv-open');
+
+    const hasExtension = href.toLowerCase().endsWith('.csv') || href.toLowerCase().endsWith('.grist');
+    let initAttribute = null;
+    if (hasExtension) {
+      initAttribute = href.toLowerCase().endsWith('.csv') ? 'initialData' : 'initialFile';
+    } else {
+      initAttribute = refElement.hasAttribute('data-initial-file') || refElement.hasAttribute('data-grist-doc-open') ? 'initialFile' :
+                      refElement.hasAttribute('data-initial-data') || refElement.hasAttribute('data-grist-csv-open') ? 'initialData' :
+                      null;
+    }
+    console.assert(initAttribute, 'Must provide initialFile, initialData or initialContent');
+
+    // Loader is shown by default, needs an explicit false to disable.
+    const loader = refElement.getAttribute('data-loader') === 'false' ? false : true;
+    previewInGrist({[initAttribute]: href, name, singlePage, loader});
+  }
+
+  window.previewInGrist = previewInGrist;
+
+  // ##################### Default styles for buttons #####################
+
+
+  style(`
+    button.grist, button.grist-large {
+      outline: none;
+      user-select: none;
+      letter-spacing: -0.08px;
+      font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Liberation Sans", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
+      color: white;
+      background: #16b378;
+      border: #16b378;
+      cursor: pointer;
+      padding: 4px 8px;
+      border-radius: 4px;
+    }
+    button.grist:hover, button.grist-large:hover {
+      background: #009058;
+      border: #009058;
+    }
+    button.grist-large {
+      font-weight: bold;
+      padding: 8px 16px;
+      min-height: 40px;
+    }
+  `);
+
+  function style(content) {
+    const styleElement = document.createElement('style');
+    styleElement.textContent = content;
+    document.head.appendChild(styleElement);
+    return styleElement;
+  }
+
+
+  // ##################### CSV drop #####################
+
+
+  class CsvDrop extends HTMLElement {
```

Comment:
This will be removed before merging. It is just a reference on how to connect those two parts.

Reply:
Got it, thanks. It would be fine to keep in a `page/` file as a reference, if you like